lib/propolis/src/tasks.rs

```diff
@@ -412,3 +416,94 @@ impl Drop for CtrlHeld<'_> {
         }
     }
 }
+
+/// Holds a group of tokio task [task::JoinHandle]s to be later joined as a
+/// group when they have all concluded.
+pub struct TaskGroup(Mutex<Vec<task::JoinHandle<()>>>);
+impl TaskGroup {
+    pub fn new() -> Self {
+        Self(Mutex::new(Vec::new()))
+    }
+
+    /// Add to the group of contained tasks
+    pub fn extend<I>(&self, tasks: I)
+    where
+        I: Iterator<Item = task::JoinHandle<()>>,
+    {
+        let mut guard = self.0.lock().unwrap();
+        guard.extend(tasks);
+    }
+
+    /// Block until all held tasks have been joined, returning any resulting
+    /// [task::JoinError]s after doing so.
+    pub fn block_until_joined(&self) -> Option<Vec<task::JoinError>> {
+        let mut guard = self.0.lock().unwrap();
+        let workers = std::mem::replace(&mut *guard, Vec::new());
+        if workers.is_empty() {
+            return None;
+        }
+
+        let rt = tokio::runtime::Handle::current();
+        let errors = rt.block_on(async {
+            FuturesUnordered::from_iter(workers)
+                .filter_map(|res| futures::future::ready(res.err()))
+                .collect::<Vec<_>>()
+                .await
+        });
+        if errors.is_empty() {
+            None
+        } else {
+            Some(errors)
+        }
+    }
+}
+
+/// Convenience type alias for the [thread::JoinHandle::join()] error type
+pub type ThreadErr = Box<dyn Any + Send + 'static>;
+
+/// Holds a group of thread [thread::JoinHandle]s to be later joined as a group
+/// when they have all concluded.
+pub struct ThreadGroup(Mutex<Vec<thread::JoinHandle<()>>>);
+impl ThreadGroup {
+    pub fn new() -> Self {
+        Self(Mutex::new(Vec::new()))
+    }
+
+    /// Add to group of contained threads
+    ///
+    /// The first (if any) [std::io::Error] encounted while among the `threads`
+    /// items, will determine the return from this function.  Despite any
+    /// encountered error, the non-error [thread::JoinHandle]s will be added to
+    /// the group.
+    pub fn extend<I>(&self, threads: I) -> std::io::Result<()>
+    where
+        I: Iterator<Item = std::io::Result<thread::JoinHandle<()>>>,
+    {
+        let mut guard = self.0.lock().unwrap();
+        let mut first_err = None;
+        for result in threads {
+            match result {
+                Err(e) => first_err = Some(e),
+                Ok(hdl) => guard.push(hdl),
+            }
+        }
+        if let Some(e) = first_err {
+            Err(e)
+        } else {
+            Ok(())
+        }
```

Comment:
style nit, take it or leave it: we could maybe simplify this to
```suggestion
        let mut first_err = Ok(());
        for result in threads {
            match result {
                Err(e) => first_err = Err(e),
                Ok(hdl) => guard.push(hdl),
            }
        }
        first_err
```

Reply:
also, is the name `first_err` really correct? isn't this actually `last_err`, since any subsequent error we see will overwrite it?

Reply:
I've reworked this to resemble your suggestion, with the logic fixed so it does emit the first error.